### PR TITLE
Define Von Neumann entropy

### PR DESCRIPTION
**Pull request by vtomole**
Part of https://github.com/quantumlib/Cirq/issues/1289

### CODE REVIEW

File: cirq/sim/density_matrix_utils.py [hunk 1]
```diff
@@ -257,3 +257,12 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
     if any(index >= num_qubits for index in indices):
         raise IndexError('Out of range indices, must be less than number of '
                          'qubits but was {}'.format(indices))
+
+
+def von_neumann_entropy(density_matrix: np.ndarray) -> np.ndarray:
+    """Calculates von Neumann entropy of density matrix"""
```
Comment (Strilanc) [inline cirq/sim/density_matrix_utils.py]: period

File: cirq/sim/density_matrix_utils.py [hunk 2]
```diff
@@ -257,3 +257,12 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
...
+    if 0 in density_matrix:
+        raise ValueError('Cannot calculate von Neumann entropy of a density '
```
Comment (Strilanc) [inline cirq/sim/density_matrix_utils.py]: Errr... I think there's been a misunderstanding. You are using per-entry arithmetic instead of whole-matrix arithmetic.

Let V be the list of eigenvalues of D. Then then von neumann entropy of D is -\sum_k V_k \log_2 V_k.

There is no constraint related to entries in the matrix having to be non-zero.

Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Ah! Yes.

Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Today i learned: https://en.wikipedia.org/wiki/Logarithm_of_a_matrix

**Comment by vtomole** [comment]
@Strilanc Why does https://www.scottaaronson.com/qclec/11.pdf (page 2) say `-\sum_k V_k 1/\log_2 V_k`? (The inverse of the eigenvalue for the log). If an eigenvalue is one, then it's a division by zero.

**Comment by vtomole** [comment]
@Strilanc PTAL

File: cirq/sim/density_matrix_utils_test.py [hunk 3]
```diff
@@ -27,12 +27,12 @@ def assert_valid_density_matrix(matrix, num_qubits=1):
...
 def test_von_neumann_entropy():
-    with pytest.raises(ValueError, match='zero'):
-        cirq.von_neumann_entropy(np.array([[1, 0], [0, 0]]))
+    assert np.isclose(cirq.von_neumann_entropy(np.array([[0.5, 0], [0, 0.5]])),
+                      -np.log(2), 1e-02)
 
     assert np.isclose(
-        cirq.von_neumann_entropy(np.array([[0.5, 0.5j], [-0.5j, 0.5]])),
-        -0.6931471805599453 + 0j)
+        cirq.von_neumann_entropy(np.array(np.array([[0.5, 0.5], [0.5, 0.5]]))),
```
Comment (Strilanc) [inline cirq/sim/density_matrix_utils_test.py]: Test the following things:

- The 1x1 matrix [[1]].

- A 2x2 density matrix created by calling random unitary, using each column as a ket, turning each ket into a density matrix, and then doing a linear combination of the two with coefficients 0.1 and 0.9. The entropy should be `-0.1 log_2(0.1) - 0.9 log_2(0.9) ~= 0.46899559358`

- An EPR pair state (|00> + |11>)(<00| + <11|). (entropy 0)

Some sort of 3x3 state.

Some sort of 4x4 state.

File: examples/examples_test.py [hunk 4]
```diff
@@ -80,4 +80,4 @@ def test_example_runs_hhl():
 
 
 def test_example_runs_qubit_characterizations():
-    examples.qubit_characterizations_example.main()
+    examples.qubit_characterizations_example.main()
```
Comment (Strilanc) [inline examples/examples_test.py]: Revert this change

**Comment by vtomole** [comment]
@Strilanc Github says your review comments were posted less than 20 mins ago and my last commit was 2 hours ago; some of which which have addressed your comments. Kind of strange.

File: cirq/sim/density_matrix_utils.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 from typing import List, Tuple, Type, Union
 
 import numpy as np
+from numpy import linalg as LA
```
[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Done.

File: cirq/sim/density_matrix_utils.py [hunk 6]
```diff
@@ -259,10 +260,16 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
...
+    Args:
+        density_matrix: The density matrix.
+        decimals: How many decimals to round the entropy to.
+
+    Returns:
+        The rounded entropy.
+
+    """
+    eigenvalues, eigenvectors = LA.eig(density_matrix)
+    return np.around(sum([i * np.log(i) for i in eigenvalues]), decimals) + 0
```
[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: The `+ 0` is to offset `-0` if that is the result

File: cirq/sim/density_matrix_utils.py [hunk 7]
```diff
@@ -259,10 +260,16 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
...
+    """Calculates von Neumann entropy of density matrix.
+    Args:
+        density_matrix: The density matrix.
+        decimals: How many decimals to round the entropy to.
+
+    Returns:
+        The rounded entropy.
+
+    """
+    eigenvalues, eigenvectors = LA.eig(density_matrix)
```
[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Done.

File: cirq/sim/density_matrix_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/sim/density_matrix_utils.py]: Using `abs` to handle the slightly-imaginary case will also handle this case.

**Comment by Strilanc** [comment]
The commit time may not be the same as the push time.

File: cirq/sim/density_matrix_utils.py [hunk 8]
```diff
@@ -259,10 +260,16 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
...
-        raise ValueError('Cannot calculate von Neumann entropy of a density '
-                         'matrix that contains a zero')
-    else:
-        return np.trace(density_matrix * np.log(density_matrix))
+def von_neumann_entropy(density_matrix: np.ndarray,
+                        decimals: int = 2) -> np.ndarray:
+    """Calculates von Neumann entropy of density matrix.
+    Args:
+        density_matrix: The density matrix.
+        decimals: How many decimals to round the entropy to.
```
[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Done.

File: cirq/sim/density_matrix_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Done.

[the remark this answers is not in the document]
Comment (vtomole) [reply cirq/sim/density_matrix_utils.py]: Done.

File: cirq/sim/density_matrix_utils_test.py (hunk 3 above)
Comment (vtomole) [reply cirq/sim/density_matrix_utils_test.py]: > An EPR pair state (|00> + |11>)(<00| + <11|)

I think i'm doing this right. With this i get some eigenvalues which are 0, which i can't take the log of.

```python
density_matrix = np.array([1, 0, 0, 1] * np.array([[1], [0], [0], [1]]))
eigenvalues = np.linalg.eigvals(density_matrix)
np.log2(eigenvalues) # RuntimeWarning: divide by zero encountered in log2 np.log2(eigenvalues)
```

**Comment by vtomole** [comment]
A reason never to pull request off of origin/master . I'll open a new one on a different branch later.
